include/sundials/priv/sundials_errors_impl.h

```diff
@@ -0,0 +1,228 @@
+ * code. It is subject to change without warning.
+ * !!!!!!!!!!!!!!!!!!!!!!!!! WARNING !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+ * -----------------------------------------------------------------
+ * Contains all error checking macros and private error handling API.
+ * -----------------------------------------------------------------*/
+
+#ifndef _SUNDIALS_ERRORS_IMPL_H
+#define _SUNDIALS_ERRORS_IMPL_H
+
+#include <sundials/sundials_errors.h>
+
+/* ----------------------------------------------------------------------------
+ * SUNErrHandler_ definition
+ * ---------------------------------------------------------------------------*/
+
+struct SUNErrHandler_
+{
+  SUNErrHandler previous; /* next error handler to call (singly linked-list) */
+  SUNErrHandlerFn call;
+  void* data;
+};
+
+SUNErrCode SUNErrHandler_Create(SUNErrHandlerFn eh_fn, void* eh_data,
+                                SUNErrHandler* eh_out);
+
+void SUNErrHandler_Destroy(SUNErrHandler eh);
+
+static inline void SUNHandleErr(int line, const char* func, const char* file,
+                                SUNErrCode code, SUNContext sunctx)
+{
+  sunctx->last_err = code;
+  SUNErrHandler eh = sunctx->err_handler;
+  while (eh != NULL)
+  {
+    eh->call(line, func, file, NULL, code, eh->data, sunctx);
+    eh = eh->previous;
+  }
+}
+
+static inline void SUNHandleErrWithMsg(int line, const char* func,
+                                       const char* file, const char* msg,
+                                       SUNErrCode code, SUNContext sunctx)
+{
+  sunctx->last_err = code;
+  SUNErrHandler eh = sunctx->err_handler;
+  while (eh != NULL)
+  {
+    eh->call(line, func, file, msg, code, eh->data, sunctx);
+    eh = eh->previous;
+  }
+}
+
+static inline void SUNHandleErrWithFmtMsg(int line, const char* func,
+                                          const char* file, const char* msgfmt,
+                                          SUNErrCode code, SUNContext sunctx, ...)
+{
+  size_t msglen;
+  char* msg;
+  va_list values;
+  va_start(values, sunctx);
+  msglen = (size_t)vsnprintf(NULL, (size_t)0, msgfmt, values); /* determine size
+                                                                  of buffer
+                                                                  needed */
+  msg = (char*)malloc(msglen + 1);
+  vsnprintf(msg, msglen + 1, msgfmt, values);
+  SUNHandleErrWithMsg(line, func, file, msg, code, sunctx);
+  va_end(values);
+  free(msg);
+}
+
+/* ----------------------------------------------------------------------------
+ * Error checking macros
+ * ---------------------------------------------------------------------------*/
+
+#define SUNCTX_ sunctx_local_scope_
+
+/* The SUNFunctionBegin macro is used to declare the SUNContext
+   object to be used a function. */
+#define SUNFunctionBegin(sunctx) \
+  SUNContext SUNCTX_ = sunctx;   \
+  (void)SUNCTX_
+
+/* SUNCheck evaluates the given expression and calls the error handler if it is
+ * false. */
```

Comment:
Point taken, but do we want the details in the developer docs or inside the header itself?

Reply:
I don't have a strong preference on this, so long as it is clear where I can point my new postdocs to read/learn.  If the internal header file documentation is expanded, then at a minimum the docs should explicitly state that developers should look there for additional documentation on other auxiliary error-handling routines.

Reply:
I think putting the documentation in the header is good. This would be consistent with how things are currently done elsewhere for internal functions. In the future we should look into using something like Breathe to include these in the developer guide as well.

Reply:
Done